librad/src/git/p2p/server.rs

```diff
@@ -70,6 +71,10 @@ impl GitServer {
             Ok(Header { service, repo, .. }) => match *service {
                 Service::UploadPack => {
                     tracing::info!("upload pack");
+                    if !self.has_repo(&repo).await? {
+                        tracing::error!("Client asked for non-existing repo: {}", repo);
+                        return send_err(&mut send, &format!("missing repo {}", repo)).await;
```

Comment:
There's no way for us to surface this particular error on the client side right? By the time it gets to `git::Remote::Error` we just have a string.

Reply:
Can you elaborate on what you mean a bit? I'm not sure I follow the question

Reply:
I'm trying to imagine how we would use this to change `librad::git::replicate` so that it can return a more informative error when the remote doesn't have the project. As I understand thing this change will end up as a `git2::Error` value when we call `DefaultFetcher::fetch`, so if we did want to do something with this particular error case we would need to match on the string in `git2::Error`.

Reply:
Well this needs to talk over the wire so I think it'd be pretty tricky to send something over and parse it on the other end. Probably not worth the hassle and just let it fail. I don't think there's much we can do if the fetcher does fail in this scenario anyway, right?

Reply:
That's fair, and it's consistent with the way that the pack protocol returns an error if the client issues a want for an object it doesn't recognise, so it probably is the right thing to do.

It seems unsatisfying that we have no way of letting the client know that the project didn't exist on the remote. I feel that such information would be helpful in debugging misbehaving servers (e.g a peer who advertises a project, then loses the project before we request it from them). I guess in such cases we'll still see the fetch errors and that's probably sufficient.

Reply:
At least we don't lose refs this way :upside_down_face:

Reply:
The problem is due to libgit2 bailing with “Early EOF” or something, instead of passing the error through. I consider it a bug.